src/compiler/checker.ts

```diff
@@ -49585,7 +49590,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     return grammarErrorOnNode(parameter.name, Diagnostics.A_rest_parameter_cannot_have_an_initializer);
                 }
             }
-            else if (isOptionalParameter(parameter)) {
+            else if (hasEffectiveQuestionToken(parameter)) {

```

Comment:
`isOptionalParameter` calls `getMinArgumentCount` for parameters with initializers and that in turn calls `getTypeOfSymbol` on rest parameters. This led to `anyType` being returned by `reportCircularityError` since the circularity was caused by this `getTypeOfSymbol` called again for something that was just being computed through out-of-order request

I spent some time on this and to the best of my knowledge - this particular function (`checkGrammarParameterList`) doesn't even have to consult that type. A parameter with an initializer is always treated as required (unless it's last) - and the main purpose of checking this here is to set `seenOptionalParameter` to potentially error on a required parameter that follows an optional one. This still happens but this error is mainly based on the syntactic~ optionality so it should be OK to just call `hasEffectiveQuestionToken` here.

I think that avoiding this extra type resolution from a grammar check function should be preferred - but I didn't recheck with all of the other existing grammar checks to see if this is a hard rule today or not.

Reply:
checkGrammar* should in general not get types. So I think this is an improvement.

> A parameter with an initializer is always treated as required

Shouldn't it be treated as optional by default?

Reply:
> Shouldn't it be treated as optional by default?

I worded this terribly here. I think I've meant that it's always treated as required when followed by a required parameter:
```ts
function fn1(a: string, b = '', c = 0) {} // b is optional
function fn2(a: string, b = '', ...args: any[]) {} // b is optional
function fn3(a: string, b = '', c: number) {} // b is required
```